### DISCUSSION TITLE
Discussion: conventions for self-merges

### DISCUSSION THREAD

**Discussion by bryevdv** [bokeh/bokeh#12423] [Governance]
We have never had any explicitly stated conventions around merges, specifically who performs them and when. I think it would be beneficial to codify some answers to these questions so that newcomers to @bokeh/core may have clear expectations. There is a wide range of possibilities, ranging from "no-self merges allowed" to "anything goes". I believe both of these are too extreme for our needs. 

A common convention I have encountered in all of my other work outside Bokeh in the last few years is: "PR authors are responsible for their own merges, once an approval is given by another team member." Given the small size and geographic distribution of the team, I think this is still (slightly) too restrictive for our situation. 

I think starting from a model of unanimous consent would keep things flexible enough for us. But that assumes that everyone is given an opportunity to raise concerns beforehand. For example, for my own PRs I often state my intentions with a time box:

> I intend to merge this PR tomorrow morning (PST) unless there is comment

I have done this as a courtesy, but I would like to propose that it become the expected norm:

* "one approval or expired time box" is the default expectation before self-merges, and
* self-merges are the default expectation for core team members (and *only* self-merges, unless explicitly asked by a PR author)

The other side of this is that if objections *are* raised, that they be addressed&mdash;either by discussion and agreement (ideally) or decision (if necessary). I believe that we owe each other due consideration of concerns. 

We can certainly carve out reasonable exceptions, e.g. for build, CI, or infra related work, for docs updates (potentially), or for "emergency" situations (credential spills, etc), or anything that might reasonably be considered "trivial". In general, I think everyone can be expected to rely on judgment here. Special circumstances may always arise, but I'd like to set a baseline for standard workflows. 

cc @bokeh/core for thoughts

**Comment by bryevdv** [comment]
Note: I am not proposing this as a BEP since it is not really related to anything directly user-facing. I would imagine this information  going in the dev guide and also potentially a Wiki page for a new contributor "onramp"

**Comment by ianthomas23** [comment]
In general I am in favour of more explicit documentation of our processes and expectations. A few comments in no particular order:

- There was some discussion at the NumFOCUS Project Summit that having expectations clearly documented can reduce the barrier for entry for new contributors and maintainers.
- As a data point, Matplotlib has a rule that 2 core maintainers must positively review a PR before it can be merged, reduced to 1 for documentation-only PRs.
- It might help to clarify the use of github labels in the PR process, i.e. the transition from `status: WIP` to `status: ready`.
- Some of the expectations listed might be considered obvious, but even as a semi-regular contributor I would appreciate it being written down somewhere as I switch between a number of projects which do not have the same etiquette/expectations.

**Reply by ianthomas23** [reply]
Noting that we have information about tags, milestones, etc in BEP 1. This is outside the docs so a search within the docs does not find the BEP. Maybe we just need to emphasise the links to the BEP with some keywords, e.g. "For information about github tags and milestones see BEP 1" so that there are some relevant search targets.

**Reply by bryevdv** [reply]
I think the dev guide should definitely link to the BEPs, I think it's just been an oversight that they aren't already. We used Wiki pages for BEPs because their updates don't necessarily align with releases/docs publishes so it was desirable for them to be maintained outside the docs.

**Reply by pavithraes** [reply]
I agree with @ianthomas23 and @bryevdv that explicitly documenting processes is helpful for a lot of folks. As another example, Dask documented some "Maintainer Guidelines" recently, [here's the one about PR merges](https://docs.dask.org/en/stable/maintainers.html#pull-requests-should-be-reviewed):
> Pull requests from non-maintainers should be reviewed and approved by at least one maintainer before being merged.
> 
> Ideally, pull requests from maintainers should also be reviewed before merging. However, because reviewer bandwidth is limited, maintainers will sometimes self-merge their own pull requests if the content of the pull request is viewed as uncontroversial (e.g. typo fix). If a maintainer has a substantial pull request which hasn’t received a review, they should ping at least one team or individual maintainer who might be interested in the proposed change. If after some time there is no response and the maintainer is confident in the proposed changes, they should post a final comment along the lines of “Merging in 48 hours if no further response” and then self-merge if no further comments are left within the specified time period. Please leave at least 48 hours to allow maintainers in other timezones to respond.
> 
> No maintainer should merge a pull request that they’re not comfortable supporting in the future.

**Reply by bryevdv** [reply]
@pavithraes thank you for sharing that, I think it is very instructive, especially the rationale about bandwidth. It's very close to what I propose, my only comparisons:
* I'd propose 24 hrs timebox for us, we with the understanding that don't need a full review in that time, but just at least a comment stating "please wait for my review" if you do want/intend to review
* I think we could *start* with a timebox rather than pinging specific individuals, but interested in others' thoughts

**Reply by pavithraes** [reply]
For context, Dask chose 48hrs to account for different time zones. I do like what you're proposing here with the understanding+comment. :)

**Comment by bryevdv** [comment]
I intend and would like to create a BEP for this and then a corresponding poll on loomio but I will wait until existing BEPs are migrated to the `pm` repo.

**Reply by bryevdv** [reply]
I've changed my mind and will draft a BEP here in the main repo
